### DISCUSSION TITLE
CRAG's retrieval evaluator as a component we could actually grade before wiring it in

### DISCUSSION THREAD

**Discussion by github-actions** [fde-academy-lab/advanced-rag-lab#132] [Reading Club]
> 💬 **Tomás** · <sub>SRE</sub>

**Papers.** Asai et al., *Self-RAG: Learning to Retrieve, Generate, and Critique
through Self-Reflection* ([arXiv:2310.11511](https://arxiv.org/abs/2310.11511)), and Yan et al.,
*Corrective Retrieval Augmented Generation*
([arXiv:2401.15884](https://arxiv.org/abs/2401.15884)).

**Claims, as I understand them.** Self-RAG has the model emit critique tokens about its own
retrieval and generation, so it can decide to retrieve again, use what it has, or decline. CRAG
puts a retrieval evaluator in front of generation that grades the retrieved set as correct,
incorrect or ambiguous and acts differently in each case.

**Why I am reading them.** Our scorecard, `structural/weighted/cross/k=8`, n=243:

```
abstention_recall        0.0
abstention_f1            0.0
false_answers_on_null     36
over_refusals              0
answer_correct        0.4115
```

Thirty-six questions in the eval set cannot be answered from the corpus, and we answer all
thirty-six. Not with a hedge — with a plausible sentence and a citation that resolves.

That is the failure I would be paged for. Every other number on the scorecard is a quality
conversation; this one is the system telling a customer something untrue with a source link
attached, and it is invisible to every average-case metric we report.

My instinct is to stop answering below a support threshold and accept the refusals, this week,
and treat the paper-shaped solutions as next quarter's work. I would like to be told why that is
wrong before I argue for it.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>

**Comment by github-actions** [comment]
> 💬 **Wei** · <sub>ML engineer · shipped a production RAG</sub>

You do not need either paper for this. Threshold the top retrieval
score and refuse below it.

We shipped precisely that on a support assistant and it held for two years. Sweep the cutoff on a
held-out set, stop where the refusal rate is one you can live with, put the number in a config
and review it monthly. It is a short function, it is explainable to a customer, and it does not
add a model call to the critical path.

The papers are interesting and they are for the case where you have already done this and the
threshold is not enough.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>

**Comment by github-actions** [comment]
> 💬 **Dan** · <sub>career-switcher · 18 months in</sub>

That sounds right to me, and it is a great deal less
frightening than fine-tuning anything. Tomás, if a cutoff gets abstention recall off zero this
week, is there a reason not to do that and read the papers afterwards?

The only worry I can think of is picking the cutoff, and Wei has given a procedure for that:
sweep it and stop where the refusal rate is tolerable. That is at least something somebody can
check rather than a hunch.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>

**Comment by github-actions** [comment]
> 💬 **Marcus** · <sub>data scientist · strong statistics</sub>

There is, and it is measured. The exercise catalogue states the
result for
[EX-18](/fde-academy-lab/advanced-rag-lab/blob/main/docs/03-exercises/catalogue.md): **no
retrieval-score threshold separates answerable from unanswerable here, best F1 0.38.** Somebody
has already swept it.

The mechanism: a null question in this corpus is a plausible question about entities that
genuinely exist, so the retriever returns confident, on-topic, well-ranked chunks, none of which
contains the answer. Low scores never enter into it. Score measures topical match, and
topical match is exactly what these questions have.

Note also that `over_refusals` is currently 0, which is the other half of the trade. Every point
of abstention recall a threshold buys is paid for somewhere in the 207 answerable questions, and
nobody has priced that side. I would want the temporal slice watched in particular — it is 66
questions at 0.091 answer correctness, so whatever makes those hard is already in the
neighbourhood of whatever a score gate would catch. That last part is a hypothesis, not a
result.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>

**Comment by github-actions** [comment]
> 💬 **Sofia** · <sub>platform engineer · security-minded</sub>

Before we go further into thresholds — how many of the 36 are
unanswerable *for this persona* rather than unanswerable full stop? We have four personas and an
ACL prefilter, and if a question is only null because the evidence was filtered out, then this is
an access-control conversation and the fix is in the prefilter and the message, not in the
generator.

A system that says "I cannot find that" when it means "you are not allowed to see that" is a
different defect, and it is the one that gets escalated.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>

**Comment by github-actions** [comment]
> 🛠 **Maintainer**

Worth separating, and here they already are: the 36 are
unanswerable by construction, for everybody. The generator builds them as plausible questions the
fact graph cannot answer, and the ACL-restricted documents are a separate built-in failure mode
with its own slice. Conflating them would make the abstention metric unreadable, because the
right behaviour differs. "The answer does not exist" and "the answer exists and is not yours" are
different sentences, and a customer needs to be told which one they are getting.

On the papers, be precise about which part of Self-RAG could run here. As I read it the critique
tokens are trained into the generator rather than prompted, so the mechanism belongs to a
fine-tuned model. Our default generator is `extractive-offline`, a scorer over sentences with a
`support_threshold` and nothing to train, so this is not a method the repository is merely short
of implementing. If you have read the training setup more recently than I have and that is wrong,
say so in this thread rather than in the note.

What transfers is the **shape of the contract**. CRAG's retrieval evaluator is a separate
component with its own inputs and its own labels, so it can be built and measured on its own
before it is wired into anything, and `sufficiency_check` in
[raglab/agent.py](/fde-academy-lab/advanced-rag-lab/blob/main/raglab/agent.py) is a crude version
of that idea already. Building the evaluator and reporting its precision and recall against the
36 is a week of work that does not need a trained model.

The condition under which the papers' numbers would return: a generator you can instruct or
train, and a null base rate resembling production. Ours is 36 of 243, far higher than real
traffic, so a policy tuned on this mix will over-refuse in the field.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>

**Comment by github-actions** [comment]
> 💬 **Tomás** · <sub>SRE</sub>

Taking the correction on the threshold. Best F1 0.38 with nothing on the other side of the ledger priced buys a different failure one
quarter later, with my name on the page, and I would have called it a regression when the refusals
started arriving.

What I have actually done: `false_answers_on_null` and `over_refusals` now sit next to each other
on the dashboard, because either one alone can be driven to zero by breaking the other. And I
have written the freeze proposal up as what it is, a policy with a price we have not measured,
rather than as the safe option. Calling it the safe option was the part I got wrong.

Next: the CRAG-shaped evaluator as a standalone component, graded against the 36 before it is
allowed near the pipeline.

---
<sub>📎 **Worked example.** Written by the maintainers to model what a good question, a good wrong turn and a good correction look like. Not a real cohort member — see the personas register in docs/10-community/. Start your own thread rather than replying here unless you have something to add.</sub>
